File: signal-layer/src/api/screener.ts
import { api } from './client.ts'
import { templateToSnake } from './signal.ts'
import type { ConditionTemplate } from '../core/types.ts'
import type { SymbolItem } from './symbol.ts'

interface ApiScreenerMatch {
  symbol: string
  name: string
  market: string
  industry?: string
  exchange?: string
  state: string
  is_ready: boolean
  progress_percent: number
}

interface ApiScreenerResponse {
  market: string
  universe_total: number
  scanned_count: number
  matched_count: number
  failed_count: number
  target_count: number
  stopped_early: boolean
  results: ApiScreenerMatch[]
}

export interface ScreenerMatch {
  item: SymbolItem
  state: string
  isReady: boolean
  progressPercent: number
}

export interface ScreenerResponse {
  market: string
  universeTotal: number
  scannedCount: number
  matchedCount: number
  failedCount: number
  targetCount: number
  stoppedEarly: boolean
  results: ScreenerMatch[]
}

export async function runScreener(
  template: ConditionTemplate,
  options: { market?: string; limit?: number; offset?: number; targetCount?: number; minProgress?: number; concurrency?: number } = {},
): Promise<ScreenerResponse> {
  const raw = await api.post<ApiScreenerResponse>('/screener/run', {
    template: templateToSnake(template),
    market: options.market ?? 'stock',
    limit: options.limit ?? 500,
    offset: options.offset ?? 0,
    target_count: options.targetCount ?? 10,
    min_progress: options.minProgress ?? 1,
    concurrency: options.concurrency ?? 4,
    kline_limit: 200,
  })
  return {
    market: raw.market,
    universeTotal: raw.universe_total,
    scannedCount: raw.scanned_count,
    matchedCount: raw.matched_count,
    failedCount: raw.failed_count,
    targetCount: raw.target_count,
    stoppedEarly: raw.stopped_early,
    results: raw.results.map((result) => ({
      item: {
        symbol: result.symbol,
        name: result.name,
        market: result.market,
        industry: result.industry,
        exchange: result.exchange,
      },
      state: result.state,
      isReady: result.is_ready,
      progressPercent: result.progress_percent,
    })),
  }
}

File: signal-layer-server/app/api/screener.py
from __future__ import annotations

import asyncio
import logging

from fastapi import APIRouter, HTTPException, Depends
from app.api.security import require_permission

from app.api.deps import (
    get_chan_service,
    get_condition_service,
    get_data_service,
    get_indicator_service,
)
from app.api.symbol import get_symbol_items
from app.schemas.screener import ScreenerMatch, ScreenerRequest, ScreenerResponse
from app.services.signal_evaluation_service import evaluate_template_for_symbol

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/screener", tags=["screener"])


@router.post("/run", response_model=ScreenerResponse, dependencies=[Depends(require_permission("screener.run"))])
async def run_screener(request: ScreenerRequest):
    """用同一策略批量评估标的，并按完成度返回匹配结果。"""
    try:
        symbols, universe_total = await get_symbol_items(
            request.market,
            request.keyword,
            request.limit,
            request.offset,
        )
    except ValueError as error:
        raise HTTPException(status_code=400, detail=str(error)) from error
    except Exception as error:
        logger.error("Screener symbol loading failed: %s", error, exc_info=True)
        raise HTTPException(status_code=500, detail=f"标的列表加载失败: {error}") from error

    data_service = get_data_service()
    chan_service = get_chan_service()
    indicator_service = get_indicator_service()
    condition_service = get_condition_service()
    failed_count = 0

    async def evaluate_item(item: dict) -> ScreenerMatch | None:
        nonlocal failed_count
        try:
            signal = await evaluate_template_for_symbol(
                item["symbol"], request.template, data_service, chan_service,
                indicator_service, condition_service, kline_limit=request.kline_limit,
            )
        except Exception as error:
            failed_count += 1
            logger.warning("Screener evaluation failed for %s: %s", item.get("symbol"), error)
            return None
        if signal.state not in request.states or signal.progress_percent < request.min_progress:
            return None
        return ScreenerMatch(
            symbol=item["symbol"], name=item.get("name", item["symbol"]),
            market=item.get("market", request.market), industry=item.get("industry"),
            exchange=item.get("exchange"), state=signal.state,
            is_ready=signal.is_ready, progress_percent=signal.progress_percent,
        )

    # 按并发数分批扫描。每批完成后立即检查目标数量，达到后不再发起下一批，
    # 因此最多只会多扫描当前批次中的少量标的。
    matches: list[ScreenerMatch] = []
    scanned_count = 0
    for offset in range(0, len(symbols), request.concurrency):
        batch = symbols[offset:offset + request.concurrency]
        evaluated = await asyncio.gather(*(evaluate_item(item) for item in batch))
        scanned_count += len(batch)
        matches.extend(item for item in evaluated if item is not None)
        if len(matches) >= request.target_count:
            break
    matches.sort(key=lambda item: (not item.is_ready, -item.progress_percent, item.symbol))
    matches = matches[:request.target_count]
    return ScreenerResponse(
        market=request.market,
        universe_total=universe_total,
        scanned_count=scanned_count,
        matched_count=len(matches),
        failed_count=failed_count,
        target_count=request.target_count,
        stopped_early=scanned_count < len(symbols),
        results=matches,
    )
